pyat/at/physics/rdt.py

```diff
@@ -0,0 +1,545 @@
+from enum import Enum
+from collections.abc import Sequence
+from dataclasses import dataclass, asdict
+
+
+__all__ = ["get_rdts", "RDTType"]
+
+
+class RDTType(Enum):
+    """Enum class for RDT type"""
+
+    ALL = 0
+    FOCUSING = 1
+    COUPLING = 2
+    CHROMATIC = 3
+    GEOMETRIC1 = 4
+    GEOMETRIC2 = 5
+    TUNESHIFT = 6
+
+
+@dataclass
+class _RDT:
+    # Location
+    refpts: Sequence[int] | int = None
+    # Normal quadrupoles rdts
+    h20000: Sequence[complex] | complex = None
+    h00200: Sequence[complex] | complex = None
+    # linear coupling rdts
+    h10010: Sequence[complex] | complex = None
+    h10100: Sequence[complex] | complex = None
+    # chromatic rdts
+    h11001: Sequence[complex] | complex = None
+    h00111: Sequence[complex] | complex = None
+    h20001: Sequence[complex] | complex = None
+    h00201: Sequence[complex] | complex = None
+    h10002: Sequence[complex] | complex = None
+    # sextupole geometric rdts
+    h21000: Sequence[complex] | complex = None
+    h30000: Sequence[complex] | complex = None
+    h10110: Sequence[complex] | complex = None
+    h10020: Sequence[complex] | complex = None
+    h10200: Sequence[complex] | complex = None
+    # octupole geometric rdts
+    h22000: Sequence[complex] | complex = None
+    h11110: Sequence[complex] | complex = None
+    h00220: Sequence[complex] | complex = None
+    h31000: Sequence[complex] | complex = None
+    h40000: Sequence[complex] | complex = None
+    h20110: Sequence[complex] | complex = None
+    h11200: Sequence[complex] | complex = None
+    h20020: Sequence[complex] | complex = None
+    h20200: Sequence[complex] | complex = None
+    h00310: Sequence[complex] | complex = None
+    h00400: Sequence[complex] | complex = None
+    # Detuning
+    dnux_dJx: Sequence[float] | float = None
+    dnux_dJy: Sequence[float] | float = None
+    dnuy_dJy: Sequence[float] | float = None
+
+    def __getattr__(self, item):
+        return asdict(self)[item]
+
+    def __getitem__(self, item):
+        return asdict(self)[item]
+
+
+def _get_polynom(elem, attr, index):
+    try:
+        val = getattr(elem, attr)[index]
+        return val
+    except (IndexError, AttributeError):
+        return 0
+
+
+def _computedrivingterms(
+    s,
+    betax,
+    betay,
+    phix,
+    phiy,
+    etax,
+    a2l,
+    b2l,
+    b3l,
+    b4l,
+    tune,
+    rdttype,
+    nperiods,
+    refpt,
+    second_order,
+):
+    """
+    Original implementation from ELEGANT
+    Based on J.Bengtsson, SLS Note 9 / 97, March 7, 1997, with corrections per W.Guo (NSLS)
+    Revised to follow C.X.Wang AOP - TN - 2009 - 020 for second - order terms
+    """
+    periodicfactor = numpy.ones((9, 9))
+    rdts = {}
+
+    def pf(i, j):
+        return periodicfactor[4 + i][4 + j]
+
+    if nperiods != 1:
+        for i in range(9):
+            for j in range(9):
+                a1 = numpy.pi * 2 * (tune[0] * (i - 4) + tune[1] * (j - 4))
```

Comment:
Consider to remove the `4` offset. This is unnecessary in numpy arrays because you can use negative indexes.
This offset is probably inherited from C, right ?